crates/sui-json-rpc/src/error.rs

```diff
@@ -94,24 +100,103 @@ impl Error {
-                    );
-                    RpcError::Call(CallError::Custom(error_object))
-                }
-                _ => RpcError::Call(CallError::Failed(err.into())),
-            },
+            Error::QuorumDriverError(err) => match_quorum_driver_error(err),
             _ => RpcError::Call(CallError::Failed(self.into())),
         }
     }
 }
 
+pub fn match_quorum_driver_error(err: QuorumDriverError) -> RpcError {
+    match err {
+        QuorumDriverError::InvalidUserSignature(err) => {
+            let inner_error_str = match err {
+                // Use inner UserInputError's Display instead of SuiError::UserInputError, which renders UserInputError in debug format
+                SuiError::UserInputError { error } => error.to_string(),
+                _ => err.to_string(),
+            };
+
+            let error_message = format!("Invalid user signature: {inner_error_str}");
+
+            let error_object = ErrorObject::owned(
+                TRANSACTION_EXECUTION_CLIENT_ERROR_CODE,
+                error_message,
+                None::<()>,
+            );
+            RpcError::Call(CallError::Custom(error_object))
+        }
+        QuorumDriverError::TimeoutBeforeFinality
+        | QuorumDriverError::FailedWithTransientErrorAfterMaximumAttempts { .. } => {
+            let error_object =
+                ErrorObject::owned(TRANSIENT_ERROR_CODE, err.to_string(), None::<()>);
+            RpcError::Call(CallError::Custom(error_object))
+        }
+        QuorumDriverError::ObjectsDoubleUsed {
+            conflicting_txes,
+            retried_tx,
+            retried_tx_success,
+        } => {
+            let error_message = format!(
+                "Failed to sign transaction by a quorum of validators because of locked objects. Retried a conflicting transaction {:?}, success: {:?}",
+                retried_tx,
+                retried_tx_success
+            );
+
+            let mut new_map: BTreeMap<TransactionDigest, Vec<ObjectRef>> = BTreeMap::new();
+
+            for (digest, (pairs, _)) in conflicting_txes {
+                let mut new_vec = Vec::new();
+
+                for (_authority, obj_ref) in pairs {
+                    new_vec.push(obj_ref);
+                }
+
+                new_map.insert(digest, new_vec);
+            }
+
+            let error_object =
+                ErrorObject::owned(INVALID_PARAMS_CODE, error_message, Some(new_map));
+            RpcError::Call(CallError::Custom(error_object))
+        }
+        QuorumDriverError::NonRecoverableTransactionError { errors } => {
+            let new_errors: Vec<String> = errors
+                .into_iter()
+                .filter(|(err, _, _)| !err.is_retryable().0) // consider retryable errors as transient errors
+                .map(|(err, _, _)| {
+                    match err {
+                        // Use inner UserInputError's Display instead of SuiError::UserInputError, which renders UserInputError in debug format
```

Comment:
Would it be bad to just change the implementation of `SuiError`'s Display to not use the debug format for `UserInputError`?